### PR TITLE
Add new component image controller

### PR DESCRIPTION
**Pull request by mmorhun**
#### What does this MR do?

This MR introduces a new controller that watches all Component related PipelineRuns and if any finish successfully, then checks for a newly built component image and updates corresponding component configuration.

#### How to test this MR?

1. Create an Application
2. Create a Component of the Application
3. Wait until initial build succeeds
4. Trigger a new build via the Component webhook
5. Wait until the webhook build finishes
6. Check the Component, it should contain the new image under `Spec.Build.ContainerImage`

### CODE REVIEW

File: controllers/component_image_controller.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+					return false
+				}
+
+				if old.Status.CompletionTime != nil {
+					// Build had been finished before this update
+					return false
+				}
+
+				// Ensure the PipelineRun belongs to a Component
+				if new.ObjectMeta.Annotations[ComponentAnnotationName] == "" && new.ObjectMeta.Labels[ComponentAnnotationName] == "" {
```
Comment (gabemontero) [inline controllers/component_image_controller.go:88]: for my own education @mmorhun  I looked at https://github.com/redhat-appstudio/application-service/blob/main/gitops/generate_build.go and I see that yeah the annotation gets set at https://github.com/redhat-appstudio/application-service/blob/86571e38f52e6bbe67b945dd58b695af26ae834d/gitops/generate_build.go#L87 as part of generating the trigger template

and the label gets set at https://github.com/redhat-appstudio/application-service/blob/86571e38f52e6bbe67b945dd58b695af26ae834d/gitops/generate_build.go#L87 for the "initial pipeline run"

given those separate paths, I'm "guessing" there is a chance where a problem could occur and one of the two might not be set, and we don't want to process in that case.

Is that the rationale? And that we can't assume if the label is there, the annotation will be, where if we could assume that, we would leverage a label selector during the construction of the cache when building the manager?

File: controllers/component_image_controller.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	switch len(tasks.Items) {
+	case 0:
+		// Build task not found, it was skipped
+		log.Info(fmt.Sprintf("No new image built in '%v' PipelineRun for '%v' Comoponent", req.NamespacedName, componentKey))
+		return ctrl.Result{}, nil
+	case 1:
+		taskrun = &tasks.Items[0]
+	default:
+		// Should not happen
+		log.Info(fmt.Sprintf("Found %d build tasks for %v PipelineRun", len(tasks.Items), pipelineRun))
```
Comment (gabemontero) [inline controllers/component_image_controller.go:208]: feels like we should minimally generate an k8s event for this

*maybe* a k8s metric / alert longer term

**Review by gabemontero — COMMENTED** [review]
a question and a minor suggestion (which I would block on if you have a strong reason/opinion on it) after an initial pass @mmorhun

**Comment by yangcao77** [comment]
HAS just removed the `Spec.Build.ContainerImage` from component CR. use `spec.containerImage` instead. So if `spec.containerImage != "" && spec.source.git != nil`, it is a build image

File: controllers/suite_util_test.go [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+	pipelineRuns := &tektonapi.PipelineRunList{}
+	labelSelectors := client.ListOptions{Raw: &metav1.ListOptions{
+		LabelSelector: "build.appstudio.openshift.io/component=" + componentKey.Name,
+	}}
+	err := k8sClient.List(ctx, pipelineRuns, &labelSelectors)
+	Expect(err).ToNot(HaveOccurred())
+	return pipelineRuns
+}
+
+func deleteComponentInitialPipelienRuns(componentKey types.NamespacedName) {
```
Comment (brunoapimentel) [inline controllers/suite_util_test.go:139]: Typo in "pipelien".

File: controllers/suite_util_test.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+		apis.Condition{
+			Reason: "Failed",
+			Status: "False",
+			Type:   apis.ConditionSucceeded,
+		},
+	}
+	Expect(k8sClient.Status().Update(ctx, pipelineRun)).Should(Succeed())
+}
+
+func succeedInitialPipelienRun(componentKey types.NamespacedName) {
```
Comment (brunoapimentel) [inline controllers/suite_util_test.go:230]: Typo in "pipelien".

File: controllers/suite_util_test.go [hunk 5]
```diff
@@ -0,0 +1,317 @@
+/*
```
Comment (brunoapimentel) [inline controllers/suite_util_test.go:1]: Nice, the helper functions look a lot better in their own file. :+1:

File: controllers/component_image_controller.go (hunk 2 above)
Comment (mmorhun) [reply controllers/component_image_controller.go]: Thanks, implemented

**Review by gabemontero — APPROVED** [review]
LGTM

**Review by Michkov — APPROVED** [review]

**Pull request merged**